subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/ChildMap.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+
+public interface ChildMap<T> {
+
+    boolean isEmpty();
+
+    List<T> values();
+
+    List<Entry<T>> entries();
+
+    void visitChildren(BiConsumer<String, ? super T> visitor);
+
+    <RESULT> RESULT getNode(VfsRelativePath targetPath, CaseSensitivity caseSensitivity, GetNodeHandler<T, RESULT> handler);
+
+    interface GetNodeHandler<T, RESULT> {
+        RESULT handleAsDescendantOfChild(VfsRelativePath pathInChild, T child);
+        RESULT handleExactMatchWithChild(T child);
+        RESULT handleUnrelatedToAnyChild();
+        RESULT handleAsAncestorOfChild(String childPath, T child);
```

Comment:
I think these methods should also receive all the available information, like in `PathRelationshipHandler`. I wonder if the two things are actually the same...

Reply:
The difference between `NodeHandler` and `PathRelationshipHandler` is that `PathRelationshipHandler` can distinguish between two children having a common prefix or not. I could change `NodeHandler` in a way that `PathRelationshipHandler` would extend it. Though I think that can lead to confusion, since you could pass a `PathRelationshipHandler` in methods which take a `NodeHandler` - and that is wrong since `handleSiblingOfChild` would never be called.

What I could do is rename `NodeHandler.handleUnrelatedToAnyChild ` to `NodeHandler.handleSiblingOrUnrelatedToAnyChild`, so it is clear what the relationship of the two methods is.

With regards to adding the extra information to `NodeHandler`: This adds quite some extra parameters to each of the methods, all of which is not needed in the subclasses. So I would not add it right now.

Reply:
I'm okay with keeping the two interfaces separate, but I think we should at least pass things as parameters that are currently grabbed from the parent scope. WDYT?